### PR TITLE
Java: Add command `BITCOUNT`

### PR DESCRIPTION
**Pull request by GumpacG**
Implement https://redis.io/docs/latest/commands/bitcount/ in the java client.

### CODE REVIEW

File: java/client/src/main/java/glide/api/commands/BitmapBaseCommands.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * Supports commands and transactions for the "Bitmap Commands" group for standalone and cluster
+ * clients.
+ *
+ * @see <a href="https://redis.io/docs/latest/commands/?group=bitmap">Bitmap Commands</a>
+ */
+public interface BitmapBaseCommands {
+    /**
+     * Counts the number of set bits (population counting) in a string. Returns zero if the key is
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/BitmapBaseCommands.java:15]: Move second sentence to `@return` section or drop

File: java/client/src/main/java/glide/api/commands/BitmapBaseCommands.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    /**
+     * Counts the number of set bits (population counting) in a string. Returns zero if the key is
+     * missing as it treated as an empty string.
+     *
+     * @see <a href="https://redis.io/commands/bitcount/">redis.io</a> for details.
+     * @param key The key to count set bits of.
+     * @param start The starting offset.
+     * @param end The ending offset.
+     * @param options The index offset type. Options are BYTE index or BIT index.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/BitmapBaseCommands.java:54]: Try to make doc in our style, see #246 for example

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 3]
```diff
@@ -2550,6 +2552,56 @@ public T touch(@NonNull String[] keys) {
         return getThis();
     }
 
+    /**
+     * Counts the number of set bits (population counting) in a string of specified <code>key</code>.
+     *
+     * @see <a href="https://redis.io/commands/bitcount/">redis.io</a> for details.
+     * @param key The key to count set bits of.
+     * @return The number set bits in the string.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:2560]: Apply below too
```suggestion
     * @return Command Response - The number set bits in the string.
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 4]
```diff
@@ -2550,6 +2552,56 @@ public T touch(@NonNull String[] keys) {
         return getThis();
     }
 
+    /**
+     * Counts the number of set bits (population counting) in a string of specified <code>key</code>.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:2556]: Copy that to interface and to all methods here
```suggestion
     * Counts the number of set bits (population counting) in a string stored at <code>key</code>.
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 5]
```diff
@@ -2550,6 +2552,56 @@ public T touch(@NonNull String[] keys) {
...
+     *
+     * @see <a href="https://redis.io/commands/bitcount/">redis.io</a> for details.
+     * @param key The key to count set bits of.
+     * @param start The starting offset.
+     * @param end The ending offset.
+     * @return The number set bits in the string.
+     */
+    public T bitcount(@NonNull String key, long start, long end) {
+        ArgsArray commandArgs =
+                buildArgs(ArrayUtils.addAll(new String[] {key, Long.toString(start), Long.toString(end)}));
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:2579]: `addAll` is not needed here

File: java/client/src/main/java/glide/api/models/commands/BitcountOptions.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    BIT("BIT");
+
+    private final String redisApi;
+
+    /**
+     * Converts BitcountOptions into a String[].
+     *
+     * @return String[]
+     */
+    public String[] toArgs() {
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/commands/BitcountOptions.java:26]: It is no-op.
You can remove this method, constructor and stored string and use `toString()` instead. We already do this in another enum somewhere.

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 7]
```diff
@@ -2550,6 +2552,56 @@ public T touch(@NonNull String[] keys) {
...
+     * @param start The starting offset.
+     * @param end The ending offset.
+     * @param options The index offset type. Options are BYTE index or BIT index.
+     * @return The number set bits in the string.
+     */
+    public T bitcount(@NonNull String key, long start, long end, BitcountOptions options) {
+        ArgsArray commandArgs =
+                buildArgs(
+                        ArrayUtils.addAll(
+                                new String[] {key, Long.toString(start), Long.toString(end)}, options.toArgs()));
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:2599]: Could be simplified everywhere (in client too)
```suggestion
                buildArgs(key, Long.toString(start), Long.toString(end), options.toString()));
```

File: java/client/src/main/java/glide/api/models/commands/BitcountOptions.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+import glide.api.commands.BitmapBaseCommands;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * Optional arguments for {@link BitmapBaseCommands#bitcount(String, long, long, BitcountOptions)}.
+ *
+ * @see <a href="https://redis.io/commands/bitcount/">redis.io</a>
+ */
+@RequiredArgsConstructor
+public enum BitcountOptions {
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/commands/BitcountOptions.java:13]: Rename to ~`BitCoinOptions`~  `BitCountOptions`

File: java/client/src/main/java/glide/api/models/commands/BitcountOptions.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    /** By default, the additional arguments start and end specify a byte index */
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/commands/BitcountOptions.java:14]: Don't copy redis docs and don't describe default behaviour
As I undestand, this parameter specifies whether indices given in the command args are bit indices or byte indices. That's correct? Try to reword the doc.

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 10]
```diff
@@ -3023,4 +3024,69 @@ public void touch(BaseClient client) {
         assertEquals(2, client.touch(new String[] {key1, key2}).get());
         assertEquals(0, client.touch(new String[] {key3}).get());
     }
+
+    @SneakyThrows
+    @ParameterizedTest(autoCloseArguments = false)
+    @MethodSource("getClients")
+    public void bitcount(BaseClient client) {
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:3031]: Combine all in one test method

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java [hunk 11]
```diff
@@ -287,6 +292,9 @@ public static Object[] transactionTestResult() {
             3L, // pfcount(new String[] { hllKey1, hllKey2 });;
             OK, // pfmerge(hllKey3, new String[] {hllKey1, hllKey2})
             3L, // pfcount(new String[] { hllKey3 })
+            15L,
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/TransactionTestUtilities.java:295]: add comments here to understand which commands these results belong to
It is extremely difficult to update or fix without comments

File: java/client/src/main/java/glide/api/commands/BitmapBaseCommands.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+     * @example
+     *     <pre>{@code
+     * Long payload = client.bitcount("myKey1").get();
+     * assert payload == 2L; // Count of set bits in a string
+     * }</pre>
+     */
+    CompletableFuture<Long> bitcount(String key);
+
+    /**
+     * Counts the number of set bits (population counting) in a string. Returns zero if the key is
```
Comment (tjzhang-BQ) [inline java/client/src/main/java/glide/api/commands/BitmapBaseCommands.java:30]: reminder for the same thing as the comment above

Comment (GumpacG) [reply java/client/src/main/java/glide/api/commands/BitmapBaseCommands.java:30]: Sorry can you clarify for which comment?

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 13]
```diff
@@ -3023,4 +3024,69 @@ public void touch(BaseClient client) {
...
+        String value = "foobar";
+
+        assertEquals(OK, client.set(key, value).get());
+        assertEquals(6, client.bitcount(key, 1, 1).get());
+    }
+
+    @SneakyThrows
+    @ParameterizedTest(autoCloseArguments = false)
+    @MethodSource("getClients")
+    public void bitcount_with_BYTE_indices(BaseClient client) {
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:3053]: bit/byte feature added in redis 7
Use `assume` to check redis version in tests

Comment (GumpacG) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:3053]: Thanks! Do we only check for redis version after 7.0.0? What about 6.0.0 and others?

Comment (Yury-Fridlyand) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:3053]: We check it for commands were added/modified in other versions.

Comment (GumpacG) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:3053]: Right but how come I don't see other `assume`s for version < 7.0.0. Can we assume that testing has a base version of 6.0.0?

Comment (Yury-Fridlyand) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:3053]: Yes, we officially support versions 6+ (see main readme).
Ci runs on 6 and 7

File: java/client/src/main/java/glide/api/commands/BitmapBaseCommands.java (hunk 1 above)
Comment (GumpacG) [reply java/client/src/main/java/glide/api/commands/BitmapBaseCommands.java]: Thanks! Addressed in https://github.com/Bit-Quill/glide-for-redis/pull/263/commits/2575b0125be5640b722ee685c33c9fd624ae1b00

File: java/client/src/main/java/glide/api/commands/BitmapBaseCommands.java (hunk 2 above)
Comment (GumpacG) [reply java/client/src/main/java/glide/api/commands/BitmapBaseCommands.java]: Thanks! Addressed in https://github.com/Bit-Quill/glide-for-redis/pull/263/commits/2575b0125be5640b722ee685c33c9fd624ae1b00

File: java/client/src/main/java/glide/api/models/BaseTransaction.java (hunk 3 above)
Comment (GumpacG) [reply java/client/src/main/java/glide/api/models/BaseTransaction.java]: Thanks! Addressed in https://github.com/Bit-Quill/glide-for-redis/pull/263/commits/2575b0125be5640b722ee685c33c9fd624ae1b00

File: java/client/src/main/java/glide/api/models/BaseTransaction.java (hunk 4 above)
Comment (GumpacG) [reply java/client/src/main/java/glide/api/models/BaseTransaction.java]: Thanks! Addressed in https://github.com/Bit-Quill/glide-for-redis/pull/263/commits/2575b0125be5640b722ee685c33c9fd624ae1b00

File: java/client/src/main/java/glide/api/models/commands/BitcountOptions.java (hunk 6 above)
Comment (GumpacG) [reply java/client/src/main/java/glide/api/models/commands/BitcountOptions.java]: Thanks! Addressed in https://github.com/Bit-Quill/glide-for-redis/pull/263/commits/2575b0125be5640b722ee685c33c9fd624ae1b00

File: java/client/src/main/java/glide/api/models/BaseTransaction.java (hunk 7 above)
Comment (GumpacG) [reply java/client/src/main/java/glide/api/models/BaseTransaction.java]: I think I've addressed where needed, let me know if I missed something though. Thanks! Addressed in https://github.com/Bit-Quill/glide-for-redis/pull/263/commits/2575b0125be5640b722ee685c33c9fd624ae1b00

File: java/client/src/main/java/glide/api/models/commands/BitcountOptions.java (hunk 8 above)
Comment (GumpacG) [reply java/client/src/main/java/glide/api/models/commands/BitcountOptions.java:14]: I think this should remain as `BitcountOptions` as the command is actually `Bitcount`.

File: java/client/src/main/java/glide/api/models/commands/BitcountOptions.java (hunk 9 above)
Comment (GumpacG) [reply java/client/src/main/java/glide/api/models/commands/BitcountOptions.java]: Thanks! Addressed in https://github.com/Bit-Quill/glide-for-redis/pull/263/commits/2575b0125be5640b722ee685c33c9fd624ae1b00

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 10 above)
Comment (GumpacG) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:3032]: Thanks! Addressed in https://github.com/Bit-Quill/glide-for-redis/pull/263/commits/2575b0125be5640b722ee685c33c9fd624ae1b00

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java (hunk 11 above)
Comment (GumpacG) [reply java/integTest/src/test/java/glide/TransactionTestUtilities.java]: Thanks! Addressed in https://github.com/Bit-Quill/glide-for-redis/pull/263/commits/2575b0125be5640b722ee685c33c9fd624ae1b00

File: java/client/src/main/java/glide/api/commands/BitmapBaseCommands.java (hunk 12 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/BitmapBaseCommands.java:30]: All docs should be same, please, align that (and in transaction too)

File: java/client/src/main/java/glide/api/commands/BitmapBaseCommands.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * Supports commands and transactions for the "Bitmap Commands" group for standalone and cluster
+ * clients.
+ *
+ * @see <a href="https://redis.io/docs/latest/commands/?group=bitmap">Bitmap Commands</a>
+ */
+public interface BitmapBaseCommands {
+    /**
+     * Counts the number of set bits (population counting) in a string.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/BitmapBaseCommands.java:15]:
```suggestion
     * Counts the number of set bits (population counting) in a string stored at <code>key</code>.
```

File: java/client/src/main/java/glide/api/models/commands/BitcountOptions.java (hunk 8 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/commands/BitcountOptions.java:14]: It is used in `bitpos` too. Need to rename according to the common sense.

File: java/client/src/main/java/glide/api/models/commands/BitcountOptions.java [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+ * @see <a href="https://redis.io/commands/bitcount/">redis.io</a>
+ */
+@RequiredArgsConstructor
+public enum BitcountOptions {
+    /** Specifies a byte index * */
+    BYTE("BYTE"),
+    /** Specifies a bit index */
+    BIT("BIT");
+
+    private final String redisApi;
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/commands/BitcountOptions.java:20]: You can simplify this even more
```suggestion
public enum BitcountOptions {
    /** Specifies a byte index * */
    BYTE,
    /** Specifies a bit index */
    BIT;
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java (hunk 7 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/models/BaseTransaction.java]: you can skip `addAll`, it is no-op

File: java/client/src/main/java/glide/api/BaseClient.java [hunk 16]
```diff
@@ -1090,4 +1094,27 @@ public CompletableFuture<String> pfmerge(
...
+                new String[] {key, Long.toString(start), Long.toString(end)},
+                this::handleLongResponse);
+    }
+
+    @Override
+    public CompletableFuture<Long> bitcount(
+            @NonNull String key, long start, long end, @NonNull BitcountOptions options) {
+
+        String[] arguments =
+                ArrayUtils.addAll(
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/BaseClient.java:1116]: remove `addAll`, it does nothing

**Review by Yury-Fridlyand — COMMENTED** [review]
Please fix tests

File: java/client/src/main/java/glide/api/models/commands/BitMapOptions.java [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+import glide.api.commands.BitmapBaseCommands;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * Optional arguments for {@link BitmapBaseCommands#bitcount(String, long, long, BitMapOptions)}.
+ * Specifies if start and end arguments are BYTE indices or BIT indices
+ *
+ * @see <a href="https://redis.io/commands/bitcount/">redis.io</a>
+ */
+@RequiredArgsConstructor
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/commands/BitMapOptions.java:13]: you can remove constructor too

**Review by Yury-Fridlyand — APPROVED** [review]

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java [hunk 18]
```diff
@@ -178,6 +179,10 @@ public static BaseTransaction<?> transactionTest(BaseTransaction<?> baseTransact
                 .pfmerge(hllKey3, new String[] {hllKey1, hllKey2})
                 .pfcount(new String[] {hllKey3});
 
+        baseTransaction.bitcount(key3);
+        baseTransaction.bitcount(key3, 2, 4);
+        baseTransaction.bitcount(key3, 2, 19, BitMapOptions.BIT);
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/TransactionTestUtilities.java:184]: Remove this test - it fails on redis 6.
It is ok to omit it, but if you want to test this in transaction - create a separate test cases for cluster and standalone client. Or wait for https://github.com/aws/glide-for-redis/pull/1284.

Comment (GumpacG) [reply java/integTest/src/test/java/glide/TransactionTestUtilities.java:184]: As discussed, it's better to wait for https://github.com/aws/glide-for-redis/pull/1284. Will add to PR to ignore GHA failures.

**Pull request merged**